File: prob1.py
#Time and Space: O(n)

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = {}

        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            
            z = tuple(count)

            if z in ans:
                ans[z].append(s)
            else:
                ans[z] = [s]

        result = list(ans.values())

        return result

File: prob1.java
// Time Complexity : O(N * K * log K)
// Space Complexity : O(NK)
// DIterate through each string in the array.
// Convert each string into a char array, sort it, and convert back to a string to create a unique representation for anagrams.
// Use a hash map to group strings with the same sorted representation, and return the grouped anagrams.

import java.util.*;

public class AnagramGrouping {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String sorted = new String(chars);
            
            map.putIfAbsent(sorted, new ArrayList<>());
            map.get(sorted).add(str);
        }
        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        AnagramGrouping anagramGrouping = new AnagramGrouping();
        String[] input = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> result = anagramGrouping.groupAnagrams(input);
        System.out.println(result);
    }
}
